Implement ETP=local on LGW

tssurya opened the pull request:
**- What this PR does and why is it needed**
This PR implements `ExternalTrafficPolicy=Local` type traffic flow on local gateway mode. The path would be `external->host->ovn-k8s-mp0->OVN->node-local-switch->load-balancer-DNATs->endpoint`

**- Special notes for reviewers**
Depends on https://github.com/ovn-org/ovn-kubernetes/pull/2663 getting in first.

**- How to verify it**
Tried and tested on KIND cluster. Docs have been written detailing the process: https://github.com/ovn-org/ovn-kubernetes/pull/2651/commits/6272b0c26f70fe8e9db1e77dbd524b7aaf231a2a

**- Description for the changelog**
`Implement ETP=local for ovn clustered pods in LGW mode`

tssurya commented:
oops I am hitting https://github.com/ovn-org/ovn-kubernetes/pull/2668: https://github.com/ovn-org/ovn-kubernetes/runs/4304241226?check_suite_focus=true#step:4:4371

tssurya commented:
/retest

tssurya commented:
/retest

tssurya commented:
/retest

tssurya commented:
/retest

tssurya commented:
/retest

tssurya commented:
@trozet and @astoycos : Ready for reviews. PTAL when you have time.

astoycos reviewed (changes requested):
Thanks for this! Some mostly small Nits, and questions 

I'm slightly worried that in LGW mode we're making LBs that aren't going to be used... but otherwise I think the general design makes sense 

one question that is probably out of the scope of this pr.... Can we add e2e CI coverage to ensure that there are not any stale IPTables rules left around if we hit certain races? I've seen in the past that this can happen since we don't have a proper IPtables controller

trozet commented on go-controller/pkg/node/gateway_iptables.go:
```diff
@@ -356,13 +390,20 @@ func getGatewayIPTRules(service *kapi.Service, hasLocalHostEndpoint bool) []iptR
 			if !hasLocalHostEndpoint {
-				for _, clusterIP := range clusterIPs {
-					rules = append(rules, getNodePortIPTRules(svcPort, clusterIP, svcPort.Port)...)
+				if etpLocalNonHostRules {
```
```			if !hasLocalHostEndpoint {
				for _, clusterIP := range clusterIPs {
					rules = append(rules, getNodePortIPTRules(svcPort, clusterIP, svcPort.Port)...)
				if etpLocalNonHostRules {
```

hasLocalHostEndpoint and etpLocalNonHostRules seem like overloaded variables that imply gateway mode, if the host has ovn-networked endpoints local to it, and if the policy has ETP=local. I'd rather see more bool params to this function which make the logic more clear.

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -70,16 +74,72 @@ type serviceConfig struct {
+// isHostNetworkedService returns etpHostRules & etpLocalNonHostRules booleans
+// etpHostRules is true if provided service has host-networked endpoints (given endpoints exist for the service)
+// etpLocalNonHostRules is true if provided service is of type etp=local & gwmode=local & endpoints are not host-networked
+func (npw *nodePortWatcher) isHostNetworkedService(service *kapi.Service) (bool, bool) {
```
is it possible to not use two different variables to declare host rules and non host rules depending on gw mode? I mean just return if the service has host network endpoints, then whatever consumes this function decides what to do based on  considering gateway mode?

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -70,16 +74,72 @@ type serviceConfig struct {
+// etpLocalNonHostRules indicates if etp=local + hn=true + gw=local
 // service func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bool, epHostLocal bool) {
-func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bool, epHostLocal bool) {
+func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bool, epHostLocal, etpLocalNonHostRules bool) {
```
again i think this should be separate parameters: epHostLocal, gw mode, etp=local

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -358,58 +432,73 @@ func (npw *nodePortWatcher) updateServiceInfo(index ktypes.NamespacedName, servi
-			if !npw.dpuMode {
-				addSharedGatewayIptRules(service, true)
+			if hasHostNet {
+				npw.updateServiceFlowCache(service, true, true, false)
```
yeah i think we should use constants for all of these

trozet commented on docs/design/external_traffic_policy.md:
````diff
@@ -58,6 +58,108 @@ eth0--->|breth0| -----> 172.18.0.3 OVN GR 100.64.0.4 --> join switch --> ovn_clu
 
 ```
 
+#### **Local Gateway Mode**
````
great documentation!

tssurya commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -601,7 +679,8 @@ func (npw *nodePortWatcher) UpdateEndpoints(old *kapi.Endpoints, new *kapi.Endpo
 	}
 
-	// Update rules if hasHostNetworkEndpoints status changed
+	// Update rules if hasHostNetworkEndpoints status changed. This handles the updates
```
yes true.

tssurya commented on go-controller/pkg/ovn/controller/services/load_balancer.go:
```diff
@@ -116,12 +118,11 @@ func buildServiceLBConfigs(service *v1.Service, endpointSlices []*discovery.Endp
-		// unless both of the following are true:
-		// - We're in shared gateway mode, and
-		// - Any of the endpoints are host-network
+		// unless any of the endpoints are host-network
```
yes, done! thanks for catching this tiny detail.

tssurya commented:
> Thanks for this! Some mostly small Nits, and questions
> 
> I'm slightly worried that in LGW mode we're making LBs that aren't going to be used... but otherwise I think the general design makes sense

right I understand, but the overhead of maintaining that migration piece of code for going from LGW to SGW is not worth it, like I'd have to find all the etp=local services and create their LBs when we go to SGW and then delete them when we go to LGW. The way forward from now on, is to make sure topology wise both modes remain exactly the same.
> 
> one question that is probably out of the scope of this pr.... Can we add e2e CI coverage to ensure that there are not any stale IPTables rules left around if we hit certain races? I've seen in the past that this can happen since we don't have a proper IPtables controller


hmm good point, I have opened https://github.com/ovn-org/ovn-kubernetes/issues/2693 to see if we can do that. I think we should surely have an e2e that at least checks these iptable rules get deleted properly. Like I caught a case where deleting the backends  (i.e deleting endpoints), still recreates iptable rules (inline comment).

tssurya commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -548,40 +627,41 @@ func (npw *nodePortWatcher) AddEndpoints(ep *kapi.Endpoints) {
 		defer npw.serviceInfoLock.Unlock()
 
 		delServiceRules(svcConfig.service, npw)
-		addServiceRules(svcConfig.service, etpHostRules, npw)
```
@astoycos, @trozet : Any idea why we were adding back the rules for endpoint deletion? Didn't make sense..so I removed it, let me know if there was some logic on why we did this.

astoycos replied:
@tssurya We still need this because of two reasons I can think of right now....The Service rules (reguardless of `etpHostEndpoints` state) should only be fully deleted once the service is deleted, not upon endpoint deletion 
If the service exists without endpoints (weird scenario) we should still funnel into OVN to reject.

tssurya replied:
ah ok, we just want to let the LB take care of all rejects if endpoints are not there right?

tssurya commented:
I haven't seen a full green CI in a long time :) , ETP=Local is passing on LGW:

```
e2e ingress traffic validation Validating ingress traffic 
  Should be allowed to node local cluster-networked endpoints by nodeport services with externalTrafficPolicy=local
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/e2e.go:1896
[BeforeEach] e2e ingress traffic validation
  /home/runner/go/pkg/mod/k8s.io/kubernetes@v1.22.2/test/e2e/framework/framework.go:185
STEP: Creating a kubernetes client
Dec  3 23:25:22.581: INFO: >>> kubeConfig: /home/runner/admin.conf
STEP: Building a namespace api object, basename nodeport-ingress-test
STEP: Waiting for a default service account to be provisioned in namespace
[BeforeEach] Validating ingress traffic
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/e2e.go:1788
STEP: Creating the endpoints pod, one for each worker
STEP: Creating an external container to send the traffic from
[It] Should be allowed to node local cluster-networked endpoints by nodeport services with externalTrafficPolicy=local
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/e2e.go:1896
STEP: Creating the nodeport service with externalTrafficPolicy=local
STEP: Waiting for the endpoints to pop up
Dec  3 23:25:45.543: INFO: Waiting for amount of service:nodeportsvclocal endpoints to be 3
STEP: Hitting the nodeport on ovn-control-plane and trying to reach only the local endpoint with protocol http
Dec  3 23:25:45.759: INFO: Validated local endpoint on node ovn-control-plane with address 172.18.0.4, and packet src IP 
%!(EXTRA string=172.18.0.5)STEP: Hitting the nodeport on ovn-worker and trying to reach only the local endpoint with protocol http
Dec  3 23:25:45.982: INFO: Validated local endpoint on node ovn-worker with address 172.18.0.2, and packet src IP 
%!(EXTRA string=172.18.0.5)STEP: Hitting the nodeport on ovn-worker2 and trying to reach only the local endpoint with protocol http
Dec  3 23:25:46.192: INFO: Validated local endpoint on node ovn-worker2 with address 172.18.0.3, and packet src IP 
%!(EXTRA string=172.18.0.5)STEP: Hitting the nodeport on ovn-control-plane and trying to reach only the local endpoint with protocol udp
Dec  3 23:25:46.399: INFO: Validated local endpoint on node ovn-control-plane with address 172.18.0.4, and packet src IP 
%!(EXTRA string=172.18.0.5)STEP: Hitting the nodeport on ovn-worker and trying to reach only the local endpoint with protocol udp
Dec  3 23:25:46.587: INFO: Validated local endpoint on node ovn-worker with address 172.18.0.2, and packet src IP 
%!(EXTRA string=172.18.0.5)STEP: Hitting the nodeport on ovn-worker2 and trying to reach only the local endpoint with protocol udp
Dec  3 23:25:46.820: INFO: Validated local endpoint on node ovn-worker2 with address 172.18.0.3, and packet src IP 
%!(EXTRA string=172.18.0.5)[AfterEach] Validating ingress traffic
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/e2e.go:1829
[AfterEach] e2e ingress traffic validation
  /home/runner/go/pkg/mod/k8s.io/kubernetes@v1.22.2/test/e2e/framework/framework.go:186
Dec  3 23:25:47.091: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "nodeport-ingress-test-1318" for this suite.
• [SLOW TEST:24.516 seconds]
e2e ingress traffic validation
/home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/e2e.go:1765
  Validating ingress traffic
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/e2e.go:1787
    Should be allowed to node local cluster-networked endpoints by nodeport services with externalTrafficPolicy=local
    /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/e2e.go:1896
```

tssurya commented:
I think this is ready for second round of reviews, and note I have changed how we perceive `etpNonHostNet` Rules like @trozet suggested and instead of making it overloaded with gateway mode and service type, I've moved out those checks into the end-consumer functions `gatewayIPTables` & `updateServiceOFRules`.

I have also added the docs for `host to service backed by ovn pod case for LGW` and corrected a few other things in the docs.

trozet commented on go-controller/pkg/node/gateway_iptables.go:
```diff
@@ -336,13 +370,14 @@ func recreateIPTRules(table, chain string, keepIPTRules []iptRule) {
+// If svcHasLocalHostNetEndPnt and svcTypeIsETPLocal, rule that redirects traffic to host targetPort is added.
+// If svcHasNonLocalHostNetEndPnt and svcTypeIsETPLocal in LGW mode, rule that redirects traffic to ovn-k8s-mp0 is added.
+// In all other cases, DNAT rule towards clusterIP svc is added.
+func getGatewayIPTRules(service *kapi.Service, svcHasLocalHostNetEndPnt, svcHasNonLocalHostNetEndPnt bool) []iptRule {
```
Is it necessary to have both bools like this?
svcHasLocalHostNetEndPnt, svcHasNonLocalHostNetEndPnt seem like opposites and you could just reduce to one? could just be localHostNetworkEndpoint

Local means local to the node, and not local gw right? Might be good to add that in the comment.

trozet replied:
I think i realize now that svcHasNonLocalHostNetEndPnt really means has OVN networked local endpoints and we should change the variable name to be more clear.

tssurya replied:
> I think i realize now that svcHasNonLocalHostNetEndPnt really means has OVN networked local endpoints and we should change the variable name to be more clear.

No Tim, `svcHasNonLocalHostNetEndPnt` means the endpoints are not - local-to-node-host-networked-endpoints :)
It does not imply "it has OVN networked local endpoints", since it also includes cases like ETP=local+host-networked-endpoints-on-different-nodes. I know this is confusing...

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -70,21 +74,64 @@ type serviceConfig struct {
 	etpHostRules bool
+	// Used when svc=etp:local + it has non local host networked backends in LGW.
+	// NOTE: For most part this is always the opposite of etpHostRules, however there
+	// can be a situation where there are no endpoints for a service in which case
```
what is the situation?

tssurya replied:
example, an idled service or a case where only svc is created and backends are not yet ready so 0 endpoints. Or a case where someone accidentally deleted the endpoints? Basically situations where endpoints.Subset is an empty list? 

Is it really ok to label that as "opposite of local-host-networked-endpoints" in these situations and then let the load-balancer controller take action - which it already does btw, the backend strings on the load balancer become "" (empty) so packet gets rejected?

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -70,21 +74,64 @@ type serviceConfig struct {
+	// NOTE: For most part this is always the opposite of etpHostRules, however there
+	// can be a situation where there are no endpoints for a service in which case
+	// both etpHostRules & etpNonHostRules will be false.
+	etpNonHostRules bool
```
would it be better to call this etpLocalNoHostEndpoints (and refactor line 76 etpHostRules)?

trozet replied:
or does this really mean etpLocalAndOVNLocalEndpoints?

tssurya replied:
yes then we'll also have to change the variable name before this... as in...

etpHostRules - needs to be etpLocalNodeHostEndpointsRules &
etpNonHostRules - will be etpNonLocalNodeHostEndpointsRules

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -70,21 +74,64 @@ type serviceConfig struct {
+		// NP
+		nodeportFlowsResult = append(nodeportFlowsResult,
+			// Traffic destined for ETP=local backed by OVN-K pod in LGW mode is matched on the svcNP and sent to table6.
+			fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_dst=%d, actions=ct(commit,zone=%d,table=6)",
```
wouldnt this only be the case in local gateway mode? in shared it should go to OVN right? Is this function only for local gw mode or what am I missing?

tssurya replied:
correct this function is only called for LGW.

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -70,21 +74,64 @@ type serviceConfig struct {
+				cookie, npw.ofportPhys, flowProtocol, nwDst, ipAddress, svcPort, HostNodePortCTZone),
+			fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, %s=%s, tp_src=%d, actions=ct(zone=%d,table=7)",
+				cookie, flowProtocol, nwSrc, ipAddress, svcPort, HostNodePortCTZone))
+	} else {
```
a service can have external IP as well as be nodeport right? they are not exclusive?

tssurya replied:
correct, this function is called is called from both `updateServiceFlowCache` and then later from `createLbAndExternalSvcFlows`. When its called from the former its a NodePort service and when its called from the later, its either for externalIP or load balancer ingress and hence it will have the `ipAddress, nwDst, nwSrc` fields as well for creating the rules.

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -70,21 +74,64 @@ type serviceConfig struct {
+// traffic will be steered into LOCAL(to send it into ovn-k8s-mp0) preserving sourceIP.
+// `add` parameter indicates if the flows should exist or be removed from the cache
+// `epHostLocal` indicates if a host networked endpoint exists for this
+// `epNonHostLocal` indicates if a non-host-networked local endpoint exists for this service
```
i would change epNonHostLocal to hasOVNLocalEndpoints or something. NonHostLocal to me sounds like it could be non host and local or non host and non local.

tssurya replied:
:D yeah we probably need to have a quick chat so that I can explain this variable's meaning.

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -139,16 +186,24 @@ func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bo
+							"actions=output:%s", etpSvcOpenFlowCookie, npw.ofportPhys))
 
+					npw.ofm.updateFlowCacheEntry(key, nodeportFlows)
+				} else if isServiceTypeETPLocal && epNonHostLocal && config.Gateway.Mode == config.GatewayModeLocal {
```
i'm not sure this is correct. From what I understand this is saying: If service is ETP Local && there are ovn networked local endpoints && local gateway, then send it LOCAL. If that is not true, then it will go to the else which will add shared gw type flows to send the packet to OVN right? Is that what we want? We would want OVN to send the TCP reset if it was a TCP packet. I guess it should be fine, but should also be fine to send it into the host whether epNonHostLocal is true or not, and then the reject would happen once the local stack sends it into mp0? Might be better for consistency.

tssurya replied:
> i'm not sure this is correct. From what I understand this is saying: If service is ETP Local && there are ovn networked local endpoints && local gateway, then send it LOCAL. If that is not true, then it will go to the else which will add shared gw type flows to send the packet to OVN right? Is that what we want? 

yes that's what we want, if its shared gateway then we'd need it to go ovn and then it hits the `skip_snat=true` LB etc.


> We would want OVN to send the TCP reset if it was a TCP packet. I guess it should be fine, but should also be fine to send it into the host whether epNonHostLocal is true or not, and then the reject would happen once the local stack sends it into mp0? Might be better for consistency.

 I am not sure I follow....

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -358,58 +420,74 @@ func (npw *nodePortWatcher) updateServiceInfo(index ktypes.NamespacedName, servi
+	addFlows := true
+	// For dpu or Full mode
+	if npw != nil {
+		npw.updateServiceFlowCache(service, addFlows, svcHasLocalHostNetEndPnt, svcHasNonLocalHostNetEndPnt)
```
svcHasNonLocalHostNetEndPnt reads like has non local, host networked endpoints. I would try to to avoid using negation in the variable name here and say svcHasOVNNetEp

tssurya replied:
> svcHasNonLocalHostNetEndPnt reads like has non local, host networked endpoints. 

that's exactly what its supposed to mean :D

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -358,58 +420,74 @@ func (npw *nodePortWatcher) updateServiceInfo(index ktypes.NamespacedName, servi
-			// Always try and delete all rules here
-			delSharedGatewayIptRules(service, true)
-			delSharedGatewayIptRules(service, false)
+			// Always try and delete all rules here in full mode & in host only mode. We don't touch iptables in dpu mode.
```
nice explanation

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -650,8 +733,8 @@ func (npwipt *nodePortWatcherIptables) SyncServices(services []interface{}) {
 		}
-		// Add correct iptables rules
-		keepIPTRules = append(keepIPTRules, getGatewayIPTRules(service, false)...)
+		// Add correct iptables rules. ETP is not implemented for smart NIC mode.
```
add a TODO?

trozet commented on go-controller/pkg/ovn/controller/services/load_balancer.go:
```diff
@@ -316,6 +318,19 @@ func buildPerNodeLBs(service *v1.Service, configs []lbConfig, nodes []nodeInfo)
 						targets = switchV6Targets
 					}
 
+					if config.externalTrafficLocal && !config.isLBOrEIP {
```
a service can be EIP and nodeport. I think you need to check if the service has nodeport here instead of "not ingress or EIP". wdyt?

trozet replied:
also in the commit message it says "We are also replacing the vip on NP,EIP&LB per
node switch LB to the special masueradeIP to
support ETP=local on LGW"

which makes me think we want to add VIP for all types listed, but your logic doesn't seem to match that?

tssurya replied:
right the commit message is wrong, I am only replacing this on nodePort svc. Let me change this to reflect the suggestion you made.

tssurya commented:
/retest

tssurya commented:
/retest

tssurya commented:
```
Summarizing 1 Failure:

[Fail] Hybrid SDN Master Operations [It] cleans up a Linux node when the OVN hostsubnet annotation is removed 
/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/hybrid-overlay/pkg/controller/master_test.go:481

```
This is weird, it passes locally but flakes in CI

tssurya commented:
/retest

tssurya commented:
shard conformance failure is flake. 
```
sha256sum: WARNING: 1 computed checksum did NOT match
185
+ '[' 1 -eq 0 ']'
186
+ echo 'Check sum '\''949f81b3c30ca03a3d4effdecda04f100fa3edc07a28b19400f72ede7c5f0491'\'' does not match file ./kind'
187
Check sum '949f81b3c30ca03a3d4effdecda04f100fa3edc07a28b19400f72ede7c5f0491' does not match file ./kind
188
+ sleep 5
189
+ false
190
+ echo 'Could not download kind binary'
191
+ exit 1
192
make: *** [Makefile:11: install-kind] Error 1
193
Could not download kind binary
194
make: Leaving directory '/home/runner/work/ovn-kubernetes/ovn-kubernetes/test'
195
Error: Process completed with exit code 2.
```

trozet commented on go-controller/pkg/node/gateway_iptables.go:
```diff
@@ -355,14 +416,26 @@ func getGatewayIPTRules(service *kapi.Service, hasLocalHostEndpoint bool) []iptR
+				// DNAT traffic to masqueradeIP instead of clusterIP.
+				rules = append(rules, getNodePortIPTRules(svcPort, "", svcPort.NodePort, true)...)
+				// add a skip SNAT rule to OVN-KUBE-SNAT-MGMTPORT to preserve sourceIP for etp=local traffic.
+				rules = append(rules, getNodePortETPLocalIPTRules(svcPort, "", int32(svcPort.TargetPort.IntValue()), true)...)
```
this is confusing, you pass true here which is svcDoesNotHaveLocalHostNetEndPnt, but you are checking !svcHasLocalHostNetEndPnt. This is another reason to change the function signature to match and not use negation. Then just pass the variable to the function and do the logic there.

tssurya replied:
ack, makes sense

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -67,23 +71,67 @@ type nodePortWatcher struct {
+			fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, %s=%s, tp_src=%d, actions=ct(zone=%d,table=7)",
+				cookie, flowProtocol, nwSrc, externalIPOrLBIngressIP, svcPort, HostNodePortCTZone))
+	} else {
+		// if externalIPOrLBIngressIP is set then this service is of type NP
```
is not set

tssurya replied:
good catch

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -67,23 +71,67 @@ type nodePortWatcher struct {
+// externalIPOrLBIngressIP is either externalIP or LB.status.ingress, nwDst&nwSrc are constants used to match on the destinations based on the flowprotocol
+func (npw *nodePortWatcher) createFlowsForNonLocalHostNetEp(cookie string, svcPort int32, flowProtocol string, externalIPOrLBIngressIP, nwDst, nwSrc string) []string {
+	var nodeportFlowsResult []string
+	if len(externalIPOrLBIngressIP) != 0 {
```
much better

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -214,42 +292,54 @@ func (npw *nodePortWatcher) createLbAndExternalSvcFlows(service *kapi.Service, s
 
 		npw.ofm.updateFlowCacheEntry(key, nodeportFlows)
 
+	} else if isServiceTypeETPLocal && !hasLocalHostNetworkEp && config.Gateway.Mode == config.GatewayModeLocal {
```
this is good

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -214,42 +292,54 @@ func (npw *nodePortWatcher) createLbAndExternalSvcFlows(service *kapi.Service, s
 
+	} else if isServiceTypeETPLocal && !hasLocalHostNetworkEp && config.Gateway.Mode == config.GatewayModeLocal {
+		// case2 (see function description for details)
+		// will hit this only if 1)gatewaymode=local & 2)svc.externaltrafficpolicy=local & 3)we don't have
```
i would suggest removing lines 326-328. I'm not sure where 3a-c are coming from and I dont think it adds more confusion than clarity.

tssurya replied:
done.

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -353,62 +443,77 @@ func (npw *nodePortWatcher) updateServiceInfo(index ktypes.NamespacedName, servi
+	// full mode || dpu mode
 	if npw != nil {
-		npw.updateServiceFlowCache(service, false, false)
+		npw.updateServiceFlowCache(service, addFlows, false)
```
nit: can just pass false here

tssurya replied:
done.

trozet commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -353,62 +443,77 @@ func (npw *nodePortWatcher) updateServiceInfo(index ktypes.NamespacedName, servi
+	// For host only mode always try and delete all rules here
+	// case1: deletes the REDIRECT rules for etp=local + host-networked pods in both gwmodes
+	delGatewayIptRules(service, true)
+	// case2Orcase3a/3b:
```
typo?

tssurya replied:
oops not sure where all the spaces flew off too...

trozet commented on go-controller/pkg/ovn/controller/services/load_balancer.go:
```diff
@@ -32,6 +43,8 @@ type lbConfig struct {
 	// (see below)
 	externalTrafficLocal bool
+	// either 1) clusterIP or 2) nodePort or 3) LoadBalancerOrExternalIP (with ingress and/or externalIPs)
+	svcType ServiceType
```
this could just be hasNodePort bool. Maybe I previously steered you wrong here. i thought this value was used elsewhere in the code to make decisions, but it looks like its only used in code you added? 
```if config.externalTrafficLocal && config.svcType == ServiceTypeNodePort {```

tssurya replied:
correct, I added it specifically for that use case alone. I can rewind it.

tssurya commented on go-controller/pkg/node/gateway_iptables.go:
```diff
@@ -121,12 +122,26 @@ func getLegacySharedGatewayInitRules(chain string, proto iptables.Protocol) []ip
+// `targetPort` is the port towards which the DNAT of the nodePort service is to be added
+//     case1: (applicable only in LGW mode where ETP=local) if svcDoesNotHaveLocalHostNetEndPnt=true, targetIP=types.HostETPLocalMasqueradeIP, targetPort=svcPort.NodePort
+//     case2: if svcDoesNotHaveLocalHostNetEndPnt=false, targetIP=clusterIP, targetPort=svcPort.Port
+func getNodePortIPTRules(svcPort kapi.ServicePort, targetIP string, targetPort int32, svcDoesNotHaveLocalHostNetEndPnt bool) []iptRule {
```
done.

tssurya commented:
> thanks @tssurya !

Thank _you_ and Andrew for bearing with all those variable names and multiple iterations
